### PR TITLE
Google Authentication for PA API (CU-8679128c8)

### PR DESCRIPTION
**Pull request by micahdbak**
# Google Authentication for PA API

## BraveSensor

This PR to BraveSensor adds new PA API calls `/pa/get-google-tokens` and `/pa/get-google-payload` that wrap `googleHelpers.paGetTokens(googleAuthCode)` and `googleHelpers.paGetPayload(googleIdToken)`, respectively.
It also uses `googleHelpers.paAuthorize` instead of `clickUpHelpers.clickUpChecker` to authorize the PA API calls `/pa/create-sensor-location`, `/pa/get-sensor-clients`, and `/pa/sensor-twilio-number`.

## Test Plan

- [x] pa.js unit tests: getGooglePayload
  - [x] for a valid Google ID token
    - [x] should respond with status 200 (OK)
    - [x] should respond with JSON data aud, iss, exp, hd, email, and name, as returned by googleHelpers.paGetPayload
    - [x] should not log any information
  - [x] for an invalid Google ID token
    - [x] should respond with status 401 (Unauthorized)
    - [x] should not respond with any JSON data
    - [x] should log this unauthorized request to get Google tokens
  - [x] for no Google ID token provided
    - [x] should respond with status 401 (Unauthorized)
    - [x] should not respond with any JSON data
    - [x] should log this unauthorized request to get Google tokens
- [x] pa.js unit tests: getGoogleTokens
  - [x] for a valid Google authorization code
    - [x] should respond with status 200 (OK)
    - [x] should respond with JSON data googleAccessToken and googleIdToken as returned by googleHelpers.paGetTokens
    - [x] should not log any information
  - [x] for an invalid Google authorization code
    - [x] should respond with status 401 (Unauthorized)
    - [x] should not respond with any JSON data
    - [x] should log this unauthorized request to get Google tokens
  - [x] for no Google authorization code provided
    - [x] should respond with status 401 (Unauthorized)
    - [x] should not respond with any JSON data
    - [x] should log this unauthorized request to get Google tokens

### CODE REVIEW

File: .travis.yml
Comment (trisapeace) [inline .travis.yml:28]: Can you please also add new environment variables to `.env.example`?

File: server/pa.js
Comment (trisapeace) [inline server/pa.js:14]: To use the validator to check that the request is valid, you need to do something like:
```
  const validationErrors = Validator.validationResult(req).formatWith(helpers.formatExpressValidationErrors)
  if (validationErrors.isEmpty()) {
```

at/near the top of the function. You can see it in the other API handler functions.

File: server/package.json
Comment (trisapeace) [inline server/package.json:23]: Just leaving a comment here as a reminder to go back and update this to the version number once the brave-alert-lib part is "deployed".

File: server/routes.js
Comment (trisapeace) [inline server/routes.js:36]: Looks good!

**Review by trisapeace — COMMENTED** [review]
Looking good!

**Comment by trisapeace** [comment]
Task linked: [CU-8679128c8 Handle Google authentication in BraveSensor server for PA API](https://app.clickup.com/t/8679128c8)

File: server/.env.example
Comment (trisapeace) [inline server/.env.example:151]: Great. Thanks for repeating this comment about where to find the appropriate values here.

File: server/.env.example
Comment (trisapeace) [inline server/.env.example:117]: Good idea to note this instead of deleting it.

File: server/test/unit/apiTest/authorizeTest.js
Comment (trisapeace) [inline server/test/unit/apiTest/authorizeTest.js:18]: Good problem-solving to figure out a way to see if the `next` function is called.

A more direct way to do the same thing is to use a Sinon stub. For example:

`function authorizeNext = sandbox.stub()`

and then you'll be able to call methods like `expect(authorizeNext).to.be.called`

File: server/testingHelpers.js
Comment (trisapeace) [inline server/testingHelpers.js:157]: Cool. Good idea to factor this out for better re-use.

File: server/test/unit/apiTest/authorizeTest.js
Comment (trisapeace) [inline server/test/unit/apiTest/authorizeTest.js:24]: Haha, good point. This isn't really needed if it's just returning the same value as what was passed in, eh?

File: server/test/unit/apiTest/authorizeTest.js
Comment (trisapeace) [inline server/test/unit/apiTest/authorizeTest.js:40]: Thanks for passing `sandbox` instead of `sinon` here! That will definitely help with cleanup.

File: server/test/unit/paTest/getGooglePayloadTest.js
Comment (trisapeace) [inline server/test/unit/paTest/getGooglePayloadTest.js:28]: I don't see where you use the spy capabilities for `pa.getGooglePayload`. Am I missing something?

File: server/test/unit/paTest/getGooglePayloadTest.js
Comment (trisapeace) [inline server/test/unit/paTest/getGooglePayloadTest.js:100]: Looks like you have everything covered. Awesome!

File: server/test/unit/paTest/getGoogleTokensTest.js
Comment (trisapeace) [inline server/test/unit/paTest/getGoogleTokensTest.js:93]: Also great!

File: server/pa.js
Comment (trisapeace) [inline server/pa.js:11]: Could you please also add some checks for these validate functions? You can probably see how this has been done elsewhere in the integration tests folder.

File: server/pa.js
Comment (trisapeace) [inline server/pa.js:50]: `idToken` sounds very generic to me. Up to you, but is there a way that you can make this name more descriptive?

**Review by trisapeace — COMMENTED** [review]
Great!

File: server/test/unit/apiTest/authorizeTest.js
Comment (tmackinnon) [inline server/test/unit/apiTest/authorizeTest.js:19]: Just for my own understanding - could you explain why you removed this `authorize_next` function and replace it with `authorizezNext` ? Was it just to use camelCase?

File: server/test/unit/paTest/getGooglePayloadTest.js
Comment (tmackinnon) [inline server/test/unit/paTest/getGooglePayloadTest.js:55]: Great job! Seems like all the possibilities are covered and the tests were super clear and easy to follow.

File: server/test/unit/paTest/getGoogleTokensTest.js
Comment (tmackinnon) [inline server/test/unit/paTest/getGoogleTokensTest.js:28]: Great work on all these tests as well, all easy to understand and follow.

File: server/pa.js
Comment (tmackinnon) [inline server/pa.js:34]: This is just a comment on readability, so up to you if you want to bother updating anything. 
But it may make your code easier to read if you store `await googleHelpers.paGetPayload(req.body.idToken)` in a variable first, then pass the variable when calling `res.json()`. I think you follow this format on line 18-19 in `getGoogleTokens`. The downside is it adds an extra line, so again your call. :)

This is what I mean:
``` 
const payload = await googleHelpers.paGetPayload(req.body.idToken)
res.status(22).json(payload)
````

**Review by tmackinnon — APPROVED** [review]
This is looking great Micah! I just had a couple comments but no major changes required on my end. I'll approve pending you address any other comments from other reviewers. :)

File: server/test/unit/apiTest/authorizeTest.js
Comment (micahdbak) [reply server/test/unit/apiTest/authorizeTest.js:19]: Exactly this! The `authorize_next` function here was from an older PR made by myself, and it wasn't caught... Theresa had previously made comments that all variables, functions, etc., in the JS repositories should consistently use camelCase, and I am fixing this mistake for that reason. :)

File: server/pa.js
Comment (micahdbak) [reply server/pa.js:34]: good point!! I'll make this change.

File: server/test/unit/apiTest/authorizeTest.js
Comment (micahdbak) [reply server/test/unit/apiTest/authorizeTest.js:18]: done!

File: server/test/unit/paTest/getGooglePayloadTest.js
Comment (micahdbak) [reply server/test/unit/paTest/getGooglePayloadTest.js:28]: gotcha! Removed!

File: server/pa.js
Comment (micahdbak) [reply server/pa.js:50]: Fixed as per changes in brave-alert-lib! Anything that originated from google now has the 'google' prefix; e.g., googleIdToken, googleAccessToken, googleAuthCode.

File: server/pa.js
Comment (micahdbak) [reply server/pa.js:11]: I don't actually see any examples of testing for these functions being done in BraveSensor, and am not sure how to go about it. I'll leave this out unless someone can give me more direction on how to do this... But I have trust that the Validator object is thoroughly tested where it is defined.

**Review by ahmaddaadaa — APPROVED** [review]

File: server/test/unit/apiTest/authorizeTest.js
Comment (tmackinnon) [reply server/test/unit/apiTest/authorizeTest.js:19]: Got it! Thanks for explaining!

**Review by tmackinnon — APPROVED** [review]
Changes since my last review look good. Approved!

File: server/test/unit/apiTest/authorizeTest.js
Comment (micahdbak) [reply server/test/unit/apiTest/authorizeTest.js:39]: Yay!

File: server/pa.js
Comment (trisapeace) [reply server/pa.js]: I agree that it isn't obvious. For future reference, the validator tests can be done in the integration tests. For example, https://github.com/bravetechnologycoop/BraveSensor/blob/main/server/test/integration/dashboardTest/submitEditClientTest.js#L821-L825

**Pull request merged**
